api/grpcserver/mesh_service.go

```diff
@@ -0,0 +1,514 @@
+	log.Info("GRPC MeshService.LayerDuration")
+	return &pb.LayerDurationResponse{Duration: &pb.SimpleInt{
+		Value: uint64(s.LayerDurationSec),
+	}}, nil
+}
+
+// MaxTransactionsPerSecond returns the max number of tx per sec (a network parameter)
+func (s MeshService) MaxTransactionsPerSecond(ctx context.Context, in *pb.MaxTransactionsPerSecondRequest) (*pb.MaxTransactionsPerSecondResponse, error) {
+	log.Info("GRPC MeshService.MaxTransactionsPerSecond")
+	return &pb.MaxTransactionsPerSecondResponse{Maxtxpersecond: &pb.SimpleInt{
+		Value: uint64(s.TxsPerBlock * s.LayerAvgSize / s.LayerDurationSec),
+	}}, nil
+}
+
+// QUERIES
+
+// AccountMeshDataQuery returns account data
+func (s MeshService) AccountMeshDataQuery(ctx context.Context, in *pb.AccountMeshDataQueryRequest) (*pb.AccountMeshDataQueryResponse, error) {
+	log.Info("GRPC MeshService.AccountMeshDataQuery")
+
+	startLayer := types.LayerID(in.MinLayer)
+
+	if startLayer > s.Mesh.LatestLayer() {
+		return nil, status.Errorf(codes.InvalidArgument, "`LatestLayer` must be less than or equal to latest layer")
+	}
+	if in.Filter == nil {
+		return nil, status.Errorf(codes.InvalidArgument, "`Filter` must be provided")
+	}
+	if in.Filter.AccountId == nil {
+		return nil, status.Errorf(codes.InvalidArgument, "`Filter.AccountId` must be provided")
+	}
+	if in.Filter.AccountMeshDataFlags == uint32(pb.AccountMeshDataFlag_ACCOUNT_MESH_DATA_FLAG_UNSPECIFIED) {
+		return nil, status.Errorf(codes.InvalidArgument, "`Filter.AccountMeshDataFlags` must set at least one bitfield")
+	}
+
+	// Read the filter flags
+	filterTx := in.Filter.AccountMeshDataFlags&uint32(pb.AccountMeshDataFlag_ACCOUNT_MESH_DATA_FLAG_TRANSACTIONS) != 0
+	filterActivations := in.Filter.AccountMeshDataFlags&uint32(pb.AccountMeshDataFlag_ACCOUNT_MESH_DATA_FLAG_ACTIVATIONS) != 0
+
+	// Gather transaction data
+	addr := types.BytesToAddress(in.Filter.AccountId.Address)
+	res := &pb.AccountMeshDataQueryResponse{}
+	if filterTx {
+		meshTxIds := s.getTxIdsFromMesh(startLayer, addr)
+		mempoolTxIds := s.Mempool.GetTxIdsByAddress(addr)
+
+		// Look up full data for all unique txids
+		txs, missing := s.Mesh.GetTransactions(append(meshTxIds, mempoolTxIds...))
+
+		// TODO: Do we ever expect txs to be missing here?
+		// E.g., if this node has not synced/received them yet.
+		if len(missing) != 0 {
+			log.Error("could not find transactions %v", missing)
+			return nil, status.Errorf(codes.Internal, "error retrieving tx data")
+		}
+		for _, t := range txs {
+			res.Data = append(res.Data, &pb.AccountMeshData{
+				Datum: &pb.AccountMeshData_Transaction{
+					Transaction: convertTransaction(t),
+				},
+			})
+		}
+	}
+
+	// Gather activation data
+	if filterActivations {
+		// We have no way to look up activations by coinbase so we have no choice
+		// but to read all of them.
+		// TODO: index activations by layer (and maybe by coinbase)
+		// See https://github.com/spacemeshos/go-spacemesh/issues/2064.
+		var activations []types.ATXID
```

Comment:
I'd at least refactor to separate functions to avoid long monolithic functions, IMO